### PR TITLE
Implement modify_list and use substitute modifyList calls with it

### PR DESCRIPTION
**Pull request by thomasp85**
This PR implements a simple and performant modifyList alternative that is ~5 times faster. It will mainly speed up element rendering where modifyList was used to apply theme settings and consistently showed up in profilings as the slowest part of the ggplot2 code

### CODE REVIEW

**Comment by thomasp85** [comment]
The tests will fail until #3003 is merged

File: R/aaa-.r [hunk 1]
```diff
@@ -52,3 +52,17 @@ df_col <- function(x, name) .subset2(x, name)
 df_rows <- function(x, i) {
   new_data_frame(lapply(x, `[`, i = i))
 }
+
+# More performant modifyList without recursion
+modify_list <- function(old, new, keep_null = FALSE) {
```
[the remark this answers is not in the document]
Comment (thomasp85) [reply R/aaa-.r]: It appears not - I'll remove it...

File: R/aaa-.r [hunk 2]
```diff
@@ -52,3 +52,17 @@ df_col <- function(x, name) .subset2(x, name)
...
+  if (keep_null) {
+    for (i in names(new)) {
+      old[i] <- list(new[[i]])
+    }
+  } else {
+    for (i in names(new)) {
+      old[[i]] <- new[[i]]
```
[the remark this answers is not in the document]
Comment (thomasp85) [reply R/aaa-.r]: struggling to find one - I might just have mirrored the approach for `keep_null`without thinking too much

[the remark this answers is not in the document]
Comment (hadley) [reply R/aaa-.r]: You'll get different behaviour if the RHS has multiple elements with the same name, but that shouldn't cause problems for our uses.

[the remark this answers is not in the document]
Comment (thomasp85) [reply R/aaa-.r]: Agree - most of these reimplementations can be kept pretty simple

File: tests/testthat/test-modify-list.R [hunk 3]
```diff
@@ -0,0 +1,25 @@
+context("modify_list")
```
Comment (hadley) [inline tests/testthat/test-modify-list.R]: Test files should match R files; so you should either rename this or pull `modify_list()` into its own file.

File: tests/testthat/test-modify-list.R [hunk 4]
```diff
@@ -0,0 +1,25 @@
...
+  b = "y",
+  c = NULL,
+  d = FALSE
+)
+
+test_that("modifyList is masked", {
+  expect_error(modifyList(testlist, testappend))
+})
+
+test_that("modify_list does what it should", {
```
Comment (hadley) [inline tests/testthat/test-modify-list.R]: I think this would be clearer broken out into separate tests like "modify_list() removes NULL elements"

**Pull request merged**